rr_iarrc/src/startlight_watcher/startlight_watcher.cpp

```diff
@@ -0,0 +1,224 @@
+            foundCircles.push_back(circle);
+        }
+    }
+    return foundCircles;
+}
+
+//@note circles returned is a vector of vectors containing <x, y, radius>
+std::vector<cv::Vec3f> findCirclesFromHough(cv::Mat &binary) {
+    /// Apply the Hough Transform to find the circles
+    std::vector<cv::Vec3f> circles;
+    double dp = 1;
+    double minDist = binary.rows/8;
+    double cannyThreshold = 100;
+    double accumThreshold = 12;
+    int minRadius = 0;
+    int maxRadius = binary.rows/3;
+    cv::HoughCircles(binary, circles, CV_HOUGH_GRADIENT, dp, minDist, cannyThreshold, accumThreshold, minRadius, maxRadius);
+    return circles;
+}
+
+//For debugging purposes
+void drawCircles(cv::Mat &debug, std::vector<cv::Vec3f> circles) {
+    // Draw the circles detected
+    for( size_t i = 0; i < circles.size(); i++ ) {
+        cv::Point center(cvRound(circles[i][0]), cvRound(circles[i][1]));
+        int radius = cvRound(circles[i][2]);
+        radius += 0.8 * radius; //add an offset for easier viewing
+        // circle outline
+        cv::circle( debug, center, radius, cv::Scalar(0,255,255), 3, 8, 0 );
+     }
+}
+
+
+
+
+/*
+ * Image callback
+ * The idea: keep checking for green circles, when we find one,
+ * go back [x] frames and see if there was a red one above and nearby.
+ *
+*/
+void img_callback(const sensor_msgs::Image::ConstPtr& msg) {
+
+    if(start_detected.data) {
+        bool_pub.publish(start_detected);
+        return; // do not to all this computing if the signal has already been seen and go broadcast
+    }
+
+    //Convert msg to Mat image
+    cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
+    cv::Mat frame = cv_ptr->image;
+
+    //gather a set of frames to run on
+    for(int i = history.size()-1; i > 0; i--) {
+        history[i] = history[i - 1];
+    }
+    history[0] = frame;
+    if (!(counttostart == history.size())) {
+        counttostart++;
+        return;
+    }
+
+    cv::Mat greenImage = getGreenImage(frame);
+    std::vector<cv::Vec3f> greenCircles = findCirclesFromContours(greenImage);
+
+    if (greenCircles.size() > 0) {
+        //@note only checks farthest history. May want to check more if you have a slower camera
+        cv::Mat redImage = getRedImage(history[history.size() - 1]);
+        std::vector<cv::Vec3f> redCircles = findCirclesFromContours(redImage);
+        for (int i = 0; i < greenCircles.size(); i++) {
+            cv::Point greenCenter(static_cast<int>(greenCircles[i][0]), static_cast<int>(greenCircles[i][1]));
+            double greenRadius = greenCircles[i][2];
+
+            for (int j = 0; j < redCircles.size(); j++) {
+                cv::Point redCenter(static_cast<int>(redCircles[j][0]), static_cast<int>(redCircles[j][1]));
+                double redRadius = redCircles[j][2];
+                if (greenCenter.y > redCenter.y && //green below red
+                    std::fabs(greenRadius - redRadius) <= radiusTolerance &&
```

Comment:
I think that x, y could be converted to col, row since greenCenter.row > redCenter.row means that green is below red (instead of greenCenter.y > redCenter.y) but it isn't completely necessary to change the code to incorporate this.

Reply:
It definitely isn't intuitive at a glance, but (0,0) in the top left is the standard and the change away from cv::Point would probably clutter code. I think the comment is enough to make it clear, so I will leave this.